narwhal/primary/src/synchronizer.rs

```diff
@@ -74,6 +344,272 @@ impl Synchronizer {
+                        entry.insert(tx);
+                        rx
+                    }
+                }
+            };
+            debug!(
+                "Processing certificate {:?} suspended: missing ancestors",
+                certificate
+            );
+            return Err(DagError::Suspended(Arc::new(std::sync::Mutex::new(Some(
+                rx,
+            )))));
+        }
+
+        self.accept_certificate_internal(certificate).await
+    }
+
+    async fn accept_certificate_internal(&self, certificate: Certificate) -> DagResult<()> {
+        let digest = certificate.digest();
+
+        // This lock must be held for the scope of the function, to ensure the operations to write
+        // certificate into storage and sending the certificate to consensus is atomic. Otherwise
+        // it would be possible for certificates to be sent to consensus not in causal order,
+        // leading to inconsistencies or consensus forks.
+        let mut pending = self.inner.pending.lock().await;
+
+        // Store the certificate. Afterwards, the certificate must be sent to consensus
+        // or Narwhal needs to shutdown, to avoid insistencies certificate store and
+        // consensus dag.
+        self.inner.certificate_store.write(certificate.clone())?;
+
+        // Update metrics for accepted certificates.
+        let highest_processed_round = self
+            .inner
+            .highest_processed_round
+            .fetch_max(certificate.round(), Ordering::AcqRel)
+            .max(certificate.round());
+        let certificate_source = if self.inner.name.eq(&certificate.origin()) {
+            "own"
+        } else {
+            "other"
+        };
+        self.inner
+            .metrics
+            .highest_processed_round
+            .with_label_values(&[certificate_source])
+            .set(highest_processed_round as i64);
+        self.inner
+            .metrics
+            .certificates_processed
+            .with_label_values(&[certificate_source])
+            .inc();
+
+        // Append the certificate to the aggregator of the
+        // corresponding round.
+        if let Err(e) = self
+            .inner
+            .append_certificate_in_aggregator(certificate.clone())
+            .await
+        {
+            warn!(
+                "Failed to aggregate certificate {} for header: {}",
+                digest, e
+            );
+            return Err(DagError::ShuttingDown);
+        }
+
+        // Send it to the consensus layer.
+        if let Err(e) = self.inner.tx_new_certificates.send(certificate).await {
+            warn!(
+                "Failed to deliver certificate {} to the consensus: {}",
+                digest, e
+            );
+            return Err(DagError::ShuttingDown);
+        }
+
+        if let Some(sender) = pending.remove(&digest) {
+            // Ignore error if there is no more listeners.
+            let _ = sender.send(());
+        }
```

Comment:
Yes this is the current assumption. I agree it should be cleaned up based on GC round. I did not try to change the `pending` logic much because it will soon be refactored in another PR.